python/sbp/client/loggers/json_logger.py

```diff
@@ -81,3 +94,89 @@ def next(self):
         warnings.warn(warn, RuntimeWarning)
     self.handle.seek(0, 0)
     raise StopIteration
+
+class MultiJSONLogIterator(JSONLogIterator):
+  """
+  MultiJSONLogIterator
+
+  The :class:`MultiJSONLogIterator` is an iterator for reading JSON logs
+  of SBP data out of multiple handles.
+
+  Parameters
+  ----------
+  handler : [handles]
+    List of handles
+
+  """
+  def __init__(self, handles, dispatcher=dispatch):
+    self.handles = handles
+    self.dispatcher = dispatcher
+
+  def flush(self):
+    for handle in self.handles:
+      handle.flush()
+
+  def close(self):
+    for handle in self.handles:
+      handle.close()
+
+  def next(self):
+    """
+    Return the next record tuple from log files containing
+    JSON-serialized SBP. If an unknown SBP message type
+    is found, it'll return the raw SBP. If EOF, throws
+    exception and then returns to start of file.
+
+    Returns
+    -------
+    (float, float, object)
+      (elapsed msec since beginning of log, UTC timestamp, msg)
+
+    """
+    datas = []
+    for handle in self.handles:
+      for line in handle:
+        try:
+          datas.append(json.loads(line))
+        except ValueError:
+          warn = "Bad JSON decoding for line %s" % line
+          warnings.warn(warn, RuntimeWarning)
+
+    for data in sorted(datas, key=itemgetter('timestamp')):
+      delta = data['delta']
+      timestamp = data['timestamp']
+      item = SBP.from_json_dict(data['data'])
+      try:
+        yield (delta, timestamp, self.dispatcher(item))
+      except KeyError:
+        yield (delta, timestamp, item)
+
+    for handle in self.handles:
+      handle.seek(0, 0)
+      raise StopIteration
+
+  @staticmethod
+  def s3_handles(aws_access_key, aws_secret_key, bucket, keys):
+    """
+    Return handles for S3 objects matching keys in bucket.
+
+    Parameters
+    ----------
+    aws_access_key : string
+      AWS access key from credentials.
+    aws_secret_key : string
+      AWS secret key from credentials.
+    bucket : string
+      S3 bucket to read objects from.
+    keys : [string]
+      List of keys in bucket to iterate over.
+
+    """
+    handles = []
+    bucket = S3Connection(aws_access_key, aws_secret_key).get_bucket(bucket)
+    for key in keys:
+      for k in bucket.list(prefix=key):
+        s = k.get_contents_as_string()
```

Comment:
Oh yeah, this is far from ideal - more like a first pass at a reference implementation. There's lot of possible optimizations here - parallel downloads of keys, partial reading of files, mmap'ing in files... Not sure how applicable it will be initially with logs hopefully being in just a single key, but something to keep an eye on long term. But there are a lot of dragons around introducing optimizations here, and wanted to get something correct first.